src/mappings/exchange.ts

```diff
@@ -58,873 +72,965 @@ export function handleTokenPurchase(event: TokenPurchase): void {
+      }
+      exchange.weightedAvgPriceUSD = BigDecimal.fromString('1000000000000000000')
+        .div(oneUSDInEth)
+        .div(exchange.weightedAvgPrice)
+      userExchangeData.tokenFeesInUSD = BigDecimal.fromString('1000000000000000000')
+        .times(userExchangeData.tokenFeesPaid)
         .div(oneUSDInEth)
         .div(exchange.price)
-        .truncate(4)
-      exchange.combinedBalanceInUSD = exchange.combinedBalanceInEth.div(oneUSDInEth).truncate(4)
-    }
-    exchange.weightedAvgPriceUSD = BigDecimal.fromString('1000000000000000000')
-      .div(oneUSDInEth)
-      .div(exchange.weightedAvgPrice)
-      .truncate(4)
-    userExchangeData.tokenFeesInUSD = BigDecimal.fromString('1000000000000000000')
-      .times(userExchangeData.tokenFeesPaid)
-      .div(oneUSDInEth)
-      .div(exchange.price)
-      .truncate(4)
-  }
-
-  // ROI calculations
-  if (exchange.price.equals(BigDecimal.fromString('0'))) {
-    // do nothing. it will cause a div by zero failure
-  } else {
-    const totalTokensToEth = exchange.tokenBalance.div(exchange.price)
-    const liquidityTokensToEth = exchange.tokenLiquidity.div(exchange.price)
-    const totalBalanceValue = totalTokensToEth.plus(exchange.ethBalance)
-    const totalLiquidityValue = liquidityTokensToEth.plus(exchange.ethLiquidity)
-    if (totalLiquidityValue.equals(BigDecimal.fromString('0'))) {
-      // do nothing. it would cause a div by zero failure
-    } else {
-      exchange.ROI = totalBalanceValue.div(totalLiquidityValue).truncate(6)
     }
-  }
+    // update now that we have usd price
+    exchange.tradeVolumeUSD = exchange.tradeVolumeUSD.plus(ethAmount.times(exchange.price))
 
-  exchange.save()
-  userExchangeData.save()
+    exchange.save()
+    userExchangeData.save()
+
+    // Nov 2 2018 is 1541116800 for dayStartTimestamp and 17837 for dayID
+    // Nov 3 2018 would be 1541116800 + 86400 and 17838. And so on, for each exchange
+    const timestamp = event.block.timestamp.toI32()
+    const dayID = timestamp / 86400 // presumably this should be rounded to an int
+    const dayStartTimestamp = dayID * 86400
+    const id = event.address
+      .toHexString()
+      .concat('-')
+      .concat(BigInt.fromI32(dayID).toString())
+
+    /****** Update Global Values ******/
+    const uniswap = Uniswap.load('1')
+    uniswap.totalVolumeInEth = uniswap.totalVolumeInEth.plus(ethAmount)
+    uniswap.totalVolumeUSD = uniswap.totalVolumeUSD.plus(ethAmount.times(exchange.price.times(exchange.priceUSD)))
+    uniswap.totalTokenSells = uniswap.totalTokenSells.plus(BigInt.fromI32(1))
+    uniswap.exchangeHistoryEntityCount = uniswap.exchangeHistoryEntityCount.plus(BigInt.fromI32(1))
+    uniswap.uniswapHistoryEntityCount = uniswap.uniswapHistoryEntityCount.plus(BigInt.fromI32(1))
+    uniswap.txCount = uniswap.txCount.plus(BigInt.fromI32(1))
+    uniswap.save()
+
+    let uniswapDayData = UniswapDayData.load(dayID.toString())
```

Comment:
I would abstract this all into a helper function, similar to `createUserDataEntity` A lot of these calculations are repeated over and over again in the 4 types of events. So this will condense the code a lot. 

It might involve some unique ways of passing in arguments, can't tell right now. Try to find the happy middle where you condense enough code, but also don't make the passing of arguments too complex and making the condensed function hard to read

Reply:
Yep seemed fine with just passing in the two ints